src/storage/aggregator.cc

```diff
@@ -0,0 +1,327 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "boost/algorithm/string.hpp"
+
+#include "base/glog_wapper.h"
+#include "base/strings.h"
+#include "common/timer.h"
+#include "storage/aggregator.h"
+#include "storage/table.h"
+
+namespace openmldb {
+namespace storage {
+
+Aggregator::Aggregator(const ::openmldb::api::TableMeta& base_meta, const ::openmldb::api::TableMeta& aggr_meta,
+                       std::shared_ptr<Table> aggr_table, const uint32_t& index_pos, const std::string& aggr_col,
+                       const AggrType& aggr_type, const std::string& ts_col, WindowType window_tpye,
+                       uint32_t window_size)
+    : base_table_schema_(base_meta.column_desc()),
+      aggr_table_schema_(aggr_meta.column_desc()),
+      aggr_table_(aggr_table),
+      index_pos_(index_pos),
+      aggr_col_(aggr_col),
+      aggr_type_(aggr_type),
+      ts_col_(ts_col),
+      window_type_(window_tpye),
+      window_size_(window_size),
+      base_row_view_(base_table_schema_),
+      aggr_row_view_(aggr_table_schema_),
+      row_builder_(aggr_table_schema_) {
+    for (int i = 0; i < base_meta.column_desc().size(); i++) {
+        if (base_meta.column_desc(i).name() == aggr_col_) {
+            aggr_col_idx_ = i;
+        }
+        if (base_meta.column_desc(i).name() == ts_col_) {
+            ts_col_idx_ = i;
+        }
+    }
+    aggr_col_type_ = base_meta.column_desc(aggr_col_idx_).data_type();
+    ts_col_type_ = base_meta.column_desc(ts_col_idx_).data_type();
+    auto dimension = dimensions_.Add();
+    dimension->set_idx(0);
+}
+
+bool Aggregator::Update(const std::string& key, const std::string& row, const uint64_t& offset) {
+    base_row_view_.Reset(reinterpret_cast<int8_t*>(const_cast<char*>(row.c_str())), row.size());
+    int64_t cur_ts;
+    switch (ts_col_type_) {
+        case DataType::kBigInt: {
+            base_row_view_.GetInt64(ts_col_idx_, &cur_ts);
+            break;
+        }
+        case DataType::kTimestamp: {
+            base_row_view_.GetTimestamp(ts_col_idx_, &cur_ts);
+            break;
+        }
+        default: {
+            LOG(ERROR) << "Unsupported timestamp type: " << ts_col_type_;
+            return false;
+        }
+    }
+    if (aggr_buffer_map_.find(key) == aggr_buffer_map_.end()) {
+        aggr_buffer_map_.emplace(key, AggrBuffer{});
+    }
+    auto& aggr_buffer = aggr_buffer_map_.at(key);
+    std::lock_guard<std::mutex> lock(*aggr_buffer.mu_);
```

Comment:
the lock is too broad. Only protect the aggr_buffer update/init/clear parts. Exclude the aggr_table read/update and codec parts.